src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -471,5 +485,45 @@ private static bool IsInRangeInclusive(int ch, int start, int end)
         {
             return (uint)(ch - start) <= (uint)(end - start);
         }
+
+        /// <summary>
+        /// A scalar that represents the Unicode replacement character U+FFFD.
+        /// </summary>
+        private const int ReplacementChar = 0xFFFD;
+
+        /// <summary>
+        /// Returns <see langword="true"/> iff <paramref name="value"/> is a valid Unicode scalar
+        /// value, i.e., is in [ U+0000..U+D7FF ], inclusive; or [ U+E000..U+10FFFF ], inclusive.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsValidUnicodeScalar(uint value)
+        {
+            // By XORing the incoming value with 0xD800, surrogate code points
+            // are moved to the range [ U+0000..U+07FF ], and all valid scalar
+            // values are clustered into the single range [ U+0800..U+10FFFF ],
+            // which allows performing a single fast range check.
+
+            return IsInRangeInclusive(value ^ 0xD800U, 0x800U, 0x10FFFFU);
+        }
+
+#if netstandard
+        private static void WriteHex(int value, ref Span<char> destination, ref int written)
+        {
+            destination[written++] = (char)Int32LsbToHexDigit(value >> 12);
+            destination[written++] = (char)Int32LsbToHexDigit((int)((value >> 8) & 0xFU));
+            destination[written++] = (char)Int32LsbToHexDigit((int)((value >> 4) & 0xFU));
+            destination[written++] = (char)Int32LsbToHexDigit((int)(value & 0xFU));
+        }
```

Comment:
Rather than incrementing the ref written each time, would it be more efficient to store written into a local, use/increment that, and then store it back at the end?  Or pass it by value and return the final value from this method?